### ISSUE TITLE
Arctan2 gradient explosion for small inputs

### ISSUE THREAD

**Issue by smorad** [google/jax#18605]
### Description

Arctan2 seems to have a large gradient when the inputs are small. If training a neural network, this blows up the weights, destroying the network. I believe the gradient of `arctan2` should be clamped to 1.0.

This behavior does not seem to be present in arctan
```python
import jax
import jax.numpy as jnp
print(jax.grad(jnp.arctan2)(1e-10, 1e-10))
# Array(5.e+09, dtype=float32, weak_type=True)

print(jax.grad(jnp.arctan)(1e-10 / 1e-10))
# Array(0.5 dtype=float32, weak_type=True)
```

Related to https://github.com/google/jax/issues/15407

### What jax/jaxlib version are you using?

0.4.19

### Which accelerator(s) are you using?

GPU

### NVIDIA GPU info

```
GeForce RTX 2080 Ti
cuda==12.2.140
```

**Comment by jakevdp** [comment]
Hi - thanks for the report. It looks to me like the gradient is returning the expected result – analytically, the gradient is

$$
\frac{\partial}{\partial y}{\rm atan2}(y, x) = \frac{-y}{x^2 + y^2} \\
\frac{\partial}{\partial x}{\rm atan2}(y, x) = \frac{x}{x^2 + y^2}
$$

It's clear here that as `x` and `y` approach zero, both partial derivatives diverge.

On the other hand, the derivative of `arctan` is:
$$
\frac{\rm d}{{\rm d}x} {\rm arctan}(x) = \frac{1}{1 + x^2}
$$
so as `x` approaches zero, the derivative approaches 1.

Both seem consistent with the output you're seeing from JAX. What do you think?

**Comment by smorad** [comment]
Whoops, good point. Even though `arctan` is well-behaved, I suppose the division in a neural network is really the issue (we explicitly divide when calling `arctan(y / x)`, where as `arctan2(y, x)` divides implicitly).

I am not sure how good of an idea it is, but I can clamp the gradient from arctan to be in [-1, 1] to prevent these explosions (at the cost of slower convergence). Is this the correct way to do so?

```python

import jax
import jax.numpy as jnp

@jax.custom_vjp
def clamped_atan2(y, x):
    return jnp.arctan2(y, x)

def clamped_atan2_f(y, x):
    return clamped_atan2(y, x), (y, x)

def clamped_atan2_b(res, g):
    y, x = res
    # Compute the gradients with respect to y and x
    dy = jax.grad(jnp.arctan2, argnums=0)(y, x)
    dx = jax.grad(jnp.arctan2, argnums=1)(y, x)

    # Clip the gradients
    dy_clipped = jnp.clip(dy, a_min=-1.0, a_max=1.0)
    dx_clipped = jnp.clip(dx, a_min=-1.0, a_max=1.0)

    # Return the product of the clipped gradients and the upstream gradient
    return g * dy_clipped, g * dx_clipped

clamped_atan2.defvjp(clamped_atan2_f, clamped_atan2_b)

standard_fn = lambda y, x: jnp.arctan2(y, x**2) 
clamped_fn = lambda y, x: clamped_atan2(y, x**2) 

# Test it does the right thing when the grad is small
y = jnp.array(2.0)
x = jnp.array(1.0)
standard = standard_fn(y, x)
clamped = clamped_fn(y, x)
standard_g = jax.grad(standard_fn)(y, x) 
clamped_g = jax.grad(clamped_fn)(y, x)

print(standard, clamped) # 1.1071488 1.1071488
print(standard_g, clamped_g) # 0.2 0.2

# Test it does the right thing when the grad is big
y = jnp.array(2e-15)
x = jnp.array(1e-10)
standard = standard_fn(y, x)
clamped = clamped_fn(y, x)
standard_g = jax.grad(standard_fn)(y, x)
clamped_g = jax.grad(clamped_fn)(y, x)

print(standard, clamped) # 0.0 0.0
print(standard_g, clamped_g) # 2500000300.0 1.0
```

**Comment by jakevdp** [comment]
You haven't said much about where you're using this function, but I would be hesitant to create custom autodiff rules that are deliberately producing incorrect derivatives for the function you're evaluating. I wonder if you might have better luck choosing an alternative function which has better-behaved derivatives in the domains you're interested in?
